### PR TITLE
Add WebP support to image decoder

### PR DESCRIPTION
**Pull request by lbhm**
## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [x] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

This PR adds WebP support to the image decoder as suggested in  #3203. WebP decoding is performed using OpenCV.

#### Additional information
Affected modules and functionalities:
 - `dali/image`: Added class for WebP images and extended `image_factory.cc`
 - `dali/operators`: Added webp to `reader/loader/utils.h`

Key points relevant for the review:

- The most critical part IMO is to verify `WebpImage::PeekShapeImpl` in `webp.cc`. I gave the format references I used to decode shape information as comments. As noted in the comments, the sync code for lossy files is not where I would expect it based on my understanding of the format reference but all my tests ran through.
- Please advise if I extended the docs in the correct location(s).
- Please advise if I modified the test correctly.
- As I am a C++ novice, I welcome any feedback on coding style, inefficiencies, etc.

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/image/webp.cc [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+Image::Shape WebpImage::PeekShapeImpl(const uint8_t *encoded_buffer, size_t length) const {
+  DALI_ENFORCE(encoded_buffer);
+  DALI_ENFORCE(length >= 16);
+
+  // RIFF header has 12 bytes
+  const uint8_t *vp8_data = encoded_buffer + 12;
+  if (vp8_data[0] == 'V' && vp8_data[1] == 'P' && vp8_data[2] == '8' && vp8_data[3] == ' ') {
+    // Simple file format (lossy)
```
Comment (klecki) [inline dali/image/webp.cc:31]: As we are reading here two bytes after vp8_data+16, can you add `DALI_ENFORCE(length >= 30);` (12 + 16 + 2).

File: dali/image/webp.cc [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    // VP8 shape information starts after the sync code
+    const int64_t W = ReadValueLE<uint16_t>(vp8_data + 14) & 0x3FFF;
+    const int64_t H = ReadValueLE<uint16_t>(vp8_data + 16) & 0x3FFF;
+
+    // VP8 always uses RGB
+    return {H, W, 3};
+  } else if (vp8_data[0] == 'V' && vp8_data[1] == 'P' && vp8_data[2] == '8' && vp8_data[3] == 'L') {
+    // Simple file format (lossless)
```
Comment (klecki) [inline dali/image/webp.cc:50]: Here the same but `DALI_ENFORCE(length >= 25);` (4 bytes after offset 9, so 4 + 9 + 12).

**Review by klecki — COMMENTED** [review]
Minor typo and length checking issues, otherwise looks ok.

**Review by JanuszL — APPROVED** [review]

File: dali/operators/decoder/image_decoder.cc [hunk 3]
```diff
@@ -172,9 +172,14 @@ a dedicated hardware decoder.
 
 The output of the decoder is in *HWC* layout.
 
-Supported formats: JPG, BMP, PNG, TIFF, PNM, PPM, PGM, PBM, JPEG 2000.
+Supported formats: JPG, BMP, PNG, TIFF, PNM, PPM, PGM, PBM, JPEG 2000, WebP.
 Please note that GPU acceleration for JPEG 2000 decoding is only available for CUDA 11.
 
+.. node::
```
[the remark this answers is not in the document]
Comment (lbhm) [reply dali/operators/decoder/image_decoder.cc]: Fixed in 727f9047f9185f49a6b05421af71fbfa7f45b3bf

File: dali/image/webp.cc (hunk 1 above)
Comment (lbhm) [reply dali/image/webp.cc:31]: Fixed in 727f9047f9185f49a6b05421af71fbfa7f45b3bf

File: dali/image/webp.cc (hunk 2 above)
Comment (lbhm) [reply dali/image/webp.cc:51]: Fixed in 727f9047f9185f49a6b05421af71fbfa7f45b3bf

**Review by klecki — APPROVED** [review]

**Comment by JanuszL** [comment]
It failed now because of:
```======================================================================
ERROR: test_operator_decoders_image.test_image_decoder_consistency(<DALIImageType.ANY_DATA: 4>, 'webp', 'db/single/webp')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/opt/dali/dali/test/python/test_operator_decoders_image.py", line 249, in _testimpl_image_decoder_consistency
    compare_pipelines(img_decoder_pipe("cpu", out_type=img_out_type, files=files),
  File "/opt/dali/dali/test/python/test_utils.py", line 209, in compare_pipelines
    out1 = pipe1.run()
  File "/usr/local/lib/python3.8/dist-packages/nvidia/dali/pipeline.py", line 918, in run
    return self.outputs()
  File "/usr/local/lib/python3.8/dist-packages/nvidia/dali/pipeline.py", line 817, in outputs
    return self._outputs()
  File "/usr/local/lib/python3.8/dist-packages/nvidia/dali/pipeline.py", line 901, in _outputs
    return self._pipe.Outputs()
RuntimeError: Critical error in pipeline:
Error when executing CPU operator decoders__Image encountered:
Error in thread 0: [/opt/dali/dali/operators/decoder/host/host_decoder.cc:41] [/opt/dali/dali/image/generic_image.cc:56] Assert on "expected_shape == decoded_shape" failed: The shape of the decoded image is different than expected. Expected 427 x 640 x 4 but got 427 x 640 x 3```
Apparently when OpenCV decodes the data with `IMREAD_UNCHANGED` flags it reads only 3 channels, while in your code you assume that VP8L always has 4. I see a [thread about OpenCV and WebP discussing this](https://groups.google.com/a/webmproject.org/g/webp-discuss/c/_ocGETvcLVw?pli=1).
In the [format description](https://developers.google.com/speed/webp/docs/riff_container) I see that there is a dedicated A bit for that:
```
Alpha (L): 1 bit
Set if any of the frames of the image contain transparency information ("alpha").
```
Please double check how OpenCV threats alpha channel in WebP format and align PeekShapeImpl with what OpenCV can return if needed or adjust test data. Otherwise we will have inconstancy.

**Comment by lbhm** [comment]
Hmm this is interesting. My understanding of the lossless simple format was that it always uses RGBA. My local tests also didn't cause any errors. The docs for VP8L (lossless simple format) indicate that the alpha bit does not matter:

> The alpha_is_used bit is a hint only, and should not impact decoding. It should be set to 0 when all alpha values are 255 in the picture, and 1 otherwise.
https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification#2_riff_header

The alpha bit you were pointing to belong to the extended format VP8X which I did not implement.

However, the Google groups link you shared indicates that at least at some point the alpha channel was simply ignored by OpenCV. I will try and figure out how OpenCV currently handles WebP.

**Comment by JanuszL** [comment]
> Hmm this is interesting. My understanding of the lossless simple format was that it always uses RGBA. My local tests also didn't cause any errors. The docs for VP8L (lossless simple format) indicate that the alpha bit does not matter:
> 
> > The alpha_is_used bit is a hint only, and should not impact decoding. It should be set to 0 when all alpha values are 255 in the picture, and 1 otherwise.
> > https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification#2_riff_header
> 
> The alpha bit you were pointing to belong to the extended format VP8X which I did not implement.
> 
> However, the Google groups link you shared indicates that at least at some point the alpha channel was simply ignored by OpenCV. I will try and figure out how OpenCV currently handles WebP.

For quick testing I used Python bindings of OpenCV - `cv2.imread`.

**Comment by JanuszL** [comment]
@lbhm - can you add also add a files with alpha channel enabled to the test dataset?

**Comment by lbhm** [comment]
I have added a check for the alpha bit in case that the format is "VP8L". "VP8 " (with a space) should always be RGB and "VP8X" is not supported by the PR.
In my local tests this change worked but I fear that my tests do not quite replicate your CI pipeline. Could you run another build @JanuszL? In case it fails again, is there any way find out on which image it failed so that I could go through it with a hex editor?

**Comment by lbhm** [comment]
Good morning @JanuszL,

> @lbhm - can you add also add files with alpha channel enabled to the test dataset?

Sorry, I didn't see your comment before posting my latest reply. I added some images with transparent elements to my [DALI_extra PR](https://github.com/NVIDIA/DALI_extra/pull/62/commits/ddcf2a6a8eac3f73cfe89741be4e7fb84404125e).

**Comment by JanuszL** [comment]
> Good morning @JanuszL,
> 
> > @lbhm - can you add also add files with alpha channel enabled to the test dataset?
> 
Hi @lbhm,

> Sorry, I didn't see your comment before posting my latest reply. I added some images with transparent elements to my [DALI_extra PR](https://github.com/NVIDIA/DALI_extra/pull/62/commits/ddcf2a6a8eac3f73cfe89741be4e7fb84404125e).

Just relaunched the CI. I hope it works this time.

**Comment by JanuszL** [comment]
@lbhm - I see that the relevant test pass. I have merged test data and dependencies, updated relevant sha in your PR and relaunched the CI before the merge.

**Review by klecki — APPROVED** [review]

**Review by JanuszL — APPROVED** [review]

**Comment by JanuszL** [comment]
@lbhm - thank you very much for your contribution!

**Pull request merged**